### PR TITLE
Update gradient interface, support AbstractDifferentiation

### PR DESCRIPTION
**Pull request by lostella**
1. Remove dependency on Zygote in favor of AbstractDifferentiation (#71, #80, cc @gdalle)
2. Update the interface to gradient computation used by algorithms, see `value_and_pullback` (similar to AbstractDifferentiation `value_and_pullback_function`)
3. Update stepsize backtracking routines, to only evaluate gradients at the end (made possible by the changes in 2.) speeding things up (need to benchmark this but it’s reasonable to expect some speedup)
4. Update tests and documentation
5. Add justfile (see [just](https://github.com/casey/just))

Work in progress, opening for feedback and to take a look at CI workflows.

TODOs
- [x] expand on AD in README and docs landing page
- [x] update benchmarks code
- [ ] benchmark changes to backtracking

### CODE REVIEW

File: docs/src/guide/custom_objectives.jl [hunk 1]
```diff
@@ -12,29 +12,32 @@
...
 # Defining the proximal mapping for a custom function type requires adding a method for [`ProximalCore.prox!`](@ref).
 # 
-# To compute gradients, ProximalAlgorithms provides a fallback definition for [`ProximalCore.gradient!`](@ref), 
-# relying on [Zygote](https://github.com/FluxML/Zygote.jl) to use automatic differentiation.
-# Therefore, you can provide any (differentiable) Julia function wherever gradients need to be taken,
-# and everything will work out of the box.
+# To compute gradients, algorithms use [`ProximalAlgorithms.value_and_pullback`](@ref):
+# this relies on [AbstractDifferentiation](https://github.com/JuliaDiff/AbstractDifferentiation.jl), for automatic differentiation
+# with any of its supported backends, when functions are wrapped in [`ProximalAlgorithms.AutoDifferentiable`](@ref),
+# as the esamples below show.
```
Comment (gdalle) [inline docs/src/guide/custom_objectives.jl:18]:
```suggestion
# as the examples below show.
```

File: docs/src/guide/getting_started.jl [hunk 2]
```diff
@@ -70,7 +73,8 @@ solution, iterations = ffb(x0=ones(2), f=quadratic_cost, g=box_indicator)
 
 # We can verify the correctness of the solution by checking that the negative gradient is orthogonal to the constraints, pointing outwards:
 
--ProximalAlgorithms.gradient(quadratic_cost, solution)[1]
+v, pb = ProximalAlgorithms.value_and_pullback(quadratic_cost, solution)
+-pb()
```
Comment (gdalle) [inline docs/src/guide/getting_started.jl:77]: Maybe explain that the pullback outputs the gradient?

File: src/ProximalAlgorithms.jl [hunk 3]
```diff
@@ -1,14 +1,46 @@
...
+
+Return a tuple containing the value of `f` at `x`, and the pullback function `pb`.
+
+Function `pb`, once called, yields the gradient of `f` at `x`.
+"""
+value_and_pullback
+
+function value_and_pullback(f::AutoDifferentiable, x)
+    fx, pb = AbstractDifferentiation.value_and_pullback_function(f.backend, f.f, x)
+    return fx, () -> pb(one(fx))[1]
```
Comment (gdalle) [inline src/ProximalAlgorithms.jl:35]: Striclty speaking this is not a pullback if it takes no input. The point of a pullback is to take a cotangent and pull it back into the input space.
In this case, the cotangent is a scalar, and taking this scalar to be 1 returns the gradient, so it's okay, but the terminology might confuse autodiff people

File: src/ProximalAlgorithms.jl [hunk 4]
```diff
@@ -1,14 +1,46 @@
...
-using ProximalCore: prox, prox!, gradient, gradient!
+using ProximalCore: prox, prox!
 
 const RealOrComplex{R} = Union{R,Complex{R}}
 const Maybe{T} = Union{T,Nothing}
 
+"""
+    AutoDifferentiable(f, backend)
+
+Wrap function `f` to be auto-differentiated using `backend`.
```
Comment (gdalle) [inline src/ProximalAlgorithms.jl:13]: Specify that it's a callable struct?

File: src/utilities/ad.jl [hunk 5]
```diff
@@ -1,14 +0,0 @@
-using Zygote: pullback
-using ProximalCore
-
-struct ZygoteFunction{F}
-    f::F
-end
-
-(f::ZygoteFunction)(x) = f.f(x)
-
-function ProximalCore.gradient!(grad, f::ZygoteFunction, x)
```
Comment (gdalle) [inline src/utilities/ad.jl:10]: The diff would be much smaller if you kept your functions `gradient` and `gradient!` but made them rely on `value_and_pullback` instead, right?

File: test/utilities/test_ad.jl [hunk 6]
```diff
@@ -2,8 +2,18 @@ using Test
 using LinearAlgebra
 using ProximalOperators: NormL1
 using ProximalAlgorithms
+using Zygote
+using ReverseDiff
+using AbstractDifferentiation: ZygoteBackend, ReverseDiffBackend
+
+@testset "Autodiff backend ($B on $T)" for (T, B) in Iterators.product(
```
Comment (gdalle) [inline test/utilities/test_ad.jl:9]: Nice :)

File: src/ProximalAlgorithms.jl (hunk 3 above)
Comment (lostella) [reply src/ProximalAlgorithms.jl:35]: Yeah right, I was unsure here, thanks for pointing that out. I guess I could call it simply “closure”

File: test/utilities/test_ad.jl (hunk 6 above)
Comment (lostella) [reply test/utilities/test_ad.jl:9]: Trust, but verify!

File: src/utilities/ad.jl (hunk 5 above)
Comment (lostella) [reply src/utilities/ad.jl:10]: I guess, but I’m thinking that these primitives don’t really fit the interface of ProximalCore maybe, see https://github.com/JuliaFirstOrder/ProximalCore.jl/pull/3

File: src/ProximalAlgorithms.jl (hunk 3 above)
Comment (lostella) [reply src/ProximalAlgorithms.jl:35]: `value_and_gradient_closure` other something like that

File: docs/src/guide/getting_started.jl (hunk 2 above)
Comment (lostella) [reply docs/src/guide/getting_started.jl]: Addressed

File: src/ProximalAlgorithms.jl (hunk 3 above)
Comment (lostella) [reply src/ProximalAlgorithms.jl:37]: Addressed

File: src/ProximalAlgorithms.jl (hunk 4 above)
Comment (lostella) [reply src/ProximalAlgorithms.jl]: Addressed

**Pull request merged**